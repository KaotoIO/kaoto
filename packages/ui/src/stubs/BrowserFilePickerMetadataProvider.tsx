import { ChangeEvent, createRef, FunctionComponent, PropsWithChildren, useCallback, useRef } from 'react';
import {
  SCHEMA_FILE_ACCEPT_PATTERN,
  SCHEMA_FILE_ACCEPT_PATTERN_SOURCE_BODY,
  SCHEMA_FILE_NAME_PATTERN,
  SCHEMA_FILE_NAME_PATTERN_SOURCE_BODY,
} from '../models/datamapper';
import { IMetadataApi, MetadataContext } from '../providers';
import { readFileAsString } from './read-file-as-string';

export const BrowserFilePickerMetadataProvider: FunctionComponent<PropsWithChildren> = (props) => {
  const fileInputRef = createRef<HTMLInputElement>();
  const fileSelectionRef = useRef<{
    resolve: (files: Record<string, string>) => void;
    reject: (error: unknown) => unknown;
  }>();
  const fileContentsRef = useRef<Record<string, string>>();

  const askUserForFileSelection = useCallback(
    (
      include: string,
      _exclude?: string,
      _options?: Record<string, unknown>,
    ): Promise<string[] | string | undefined> => {
      if (!fileInputRef.current) return Promise.resolve(undefined);

      if (include === SCHEMA_FILE_NAME_PATTERN_SOURCE_BODY) {
        fileInputRef.current.accept = SCHEMA_FILE_ACCEPT_PATTERN_SOURCE_BODY;
      } else if (include === SCHEMA_FILE_NAME_PATTERN) {
        fileInputRef.current.accept = SCHEMA_FILE_ACCEPT_PATTERN;
      }

      fileInputRef.current.click();
      return new Promise<Record<string, string>>((resolve, reject) => {
        fileSelectionRef.current = { resolve, reject };
      }).then((files) => {
        fileContentsRef.current = files;
        return Object.keys(files);
      });
    },
    [fileInputRef],
  );

  const onImport = useCallback(async (event: ChangeEvent<HTMLInputElement>) => {
    const schemaFiles = event.target.files;
    if (!schemaFiles) return;
    const fileContents: Record<string, string> = {};
    const fileContentPromises: Promise<string>[] = [];
    Array.from(schemaFiles).forEach((f) => {
      const promise = readFileAsString(f).then((content) => (fileContents[f.name] = content));
      fileContentPromises.push(promise);
    });
    await Promise.allSettled(fileContentPromises);

    fileSelectionRef.current?.resolve(fileContents);
    fileSelectionRef.current = undefined;
    event.target.value = '';
  }, []);

  const getResourceContent = useCallback((path: string) => {
    return Promise.resolve(fileContentsRef.current && fileContentsRef.current[path]);
  }, []);

  const metadataApi: IMetadataApi = {
    askUserForFileSelection: askUserForFileSelection,
    getResourceContent: getResourceContent,
    shouldSaveSchema: true,
    getMetadata: () => Promise.resolve(undefined),
    setMetadata: () => Promise.resolve(),
    deleteResource: () => Promise.resolve(true),
    saveResourceContent: () => Promise.resolve(),
    getSuggestions: () => Promise.resolve([]),
    onStepUpdated: () => Promise.resolve(),
  };

  return (
    <MetadataContext.Provider value={metadataApi}>
      {props.children}
      <input
        type="file"
        style={{ display: 'none' }}
        data-testid={`attach-schema-file-input`}
        onChange={onImport}
        ref={fileInputRef}
      />
    </MetadataContext.Provider>
  );
};
