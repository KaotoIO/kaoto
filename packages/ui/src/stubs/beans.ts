export const beansYaml = `
- beans:
  - name: myBean
    type: io.kaoto.MyBean
    properties:
      p1: p1v
      p2: p2v
  - name: myBean2
    type: io.kaoto.MyBean
    properties:
      p1:
        p1s1: p1s1v
`

export const beansJson = {
  "beans": [
    {
      "name": "myBean",
      "type": "io.kaoto.MyBean",
      "properties": {
        "p1": "p1v",
        "p2": "p2v"
      }
    },
    {
      "name": "myBean2",
      "type": "io.kaoto.MyBean",
      "properties": {
        "p1": {
          "p1s1": "p1s1v"
        }
      }
    }
  ]
}

