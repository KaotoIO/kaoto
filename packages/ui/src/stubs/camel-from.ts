import { FromDefinition } from '@kaoto-next/camel-catalog/types';
import { CamelRouteVisualEntity } from '../models/visualization/flows';

/**
 * This is a stub Camel From in YAML format.
 * It is used to test the Canvas component.
 */
export const camelFromYaml = `
- from:
    uri: timer:tutorial
    steps:
    - to:
        uri: direct:my-route
`;

/**
 * This is a stub Camel From in JSON format.
 * It is used to test the Canvas component.
 */
export const camelFromJson: { from: FromDefinition } = {
  from: {
    uri: 'timer:tutorial',
    steps: [
      {
        to: {
          uri: 'direct:my-route',
        },
      },
    ],
  },
};

export const camelFrom = new CamelRouteVisualEntity({ from: camelFromJson.from });
