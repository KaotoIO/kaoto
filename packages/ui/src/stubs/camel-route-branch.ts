import { parse } from 'yaml';

export const camelRouteBranch = parse(`
- route:
    id: route-2768
    from:
      id: from-4014
      uri: timer:template
      parameters:
        period: "1000"
      steps:
        - choice:
            id: choice-3431
            otherwise:
              id: otherwise-3653
              steps:
                - log:
                    id: log-6808
                    message: \${body}
            when:
              - id: when-4112
                steps:
                  - setHeader:
                      id: setHeader-6078
                      expression:
                        simple: {}
                expression:
                  simple:
                    expression: \${header.foo} == 1
        - to:
            id: to-3757
            uri: sql
            parameters: {}
`);
