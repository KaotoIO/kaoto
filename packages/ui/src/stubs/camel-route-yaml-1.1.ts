export const camelRouteYaml_1_1_original = `
- route:
    id: route-3376
    from:
      id: from-3505
      uri: rest:post:/newCustomer
      parameters:
        host: localhost
      steps: []
`;

export const camelRouteYaml_1_1_updated = `
- route:
    id: route-3376
    from:
      id: from-3505
      uri: rest:post:/newCustomer
      parameters:
        bindingMode: "off"
        host: localhost
      steps: []
`;
