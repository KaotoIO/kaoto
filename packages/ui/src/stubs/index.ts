export * from './camel-route';
export * from './kamelet-binding-route';
export * from './kamelet-route';
export * from './pipe';
