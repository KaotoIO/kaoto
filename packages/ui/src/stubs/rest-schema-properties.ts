import { KaotoSchemaDefinition } from '../models/kaoto-schema';

export const restSchemaProperties: KaotoSchemaDefinition['schema']['properties'] = {
  id: {
    type: 'string',
    title: 'Id',
    description: 'Sets the id of this node',
  },
  description: {
    type: 'string',
    title: 'Description',
    description: 'Sets the description of this node',
  },
  disabled: {
    type: 'boolean',
    title: 'Disabled',
    description:
      'Whether to disable this REST service from the route during build time. Once an REST service has been disabled then it cannot be enabled later at runtime.',
    default: false,
  },
  path: {
    type: 'string',
    title: 'Path',
    description: 'Path of the rest service, such as /foo',
  },
  consumes: {
    type: 'string',
    title: 'Consumes',
    description:
      'To define the content type what the REST service consumes (accept as input), such as application/xml or application/json. This option will override what may be configured on a parent level',
  },
  produces: {
    type: 'string',
    title: 'Produces',
    description:
      'To define the content type what the REST service produces (uses for output), such as application/xml or application/json This option will override what may be configured on a parent level',
  },
  bindingMode: {
    type: 'string',
    title: 'Binding Mode',
    description:
      'Sets the binding mode to use. This option will override what may be configured on a parent level The default value is auto',
    default: 'off',
    enum: ['off', 'auto', 'json', 'xml', 'json_xml'],
  },
  skipBindingOnErrorCode: {
    type: 'boolean',
    title: 'Skip Binding On Error Code',
    description:
      'Whether to skip binding on output if there is a custom HTTP error code header. This allows to build custom error messages that do not bind to json / xml etc, as success messages otherwise will do. This option will override what may be configured on a parent level',
    default: false,
  },
  clientRequestValidation: {
    type: 'boolean',
    title: 'Client Request Validation',
    description:
      'Whether to enable validation of the client request to check: 1) Content-Type header matches what the Rest DSL consumes; returns HTTP Status 415 if validation error. 2) Accept header matches what the Rest DSL produces; returns HTTP Status 406 if validation error. 3) Missing required data (query parameters, HTTP headers, body); returns HTTP Status 400 if validation error. 4) Parsing error of the message body (JSon, XML or Auto binding mode must be enabled); returns HTTP Status 400 if validation error.',
    default: false,
  },
  enableCORS: {
    type: 'boolean',
    title: 'Enable CORS',
    description:
      'Whether to enable CORS headers in the HTTP response. This option will override what may be configured on a parent level The default value is false.',
    default: false,
  },
  enableNoContentResponse: {
    type: 'boolean',
    title: 'Enable No Content Response',
    description:
      'Whether to return HTTP 204 with an empty body when a response contains an empty JSON object or XML root object. The default value is false.',
    default: false,
  },
  apiDocs: {
    type: 'boolean',
    title: 'Api Docs',
    description:
      'Whether to include or exclude this rest operation in API documentation. This option will override what may be configured on a parent level. The default value is true.',
    default: true,
  },
  tag: {
    type: 'string',
    title: 'Tag',
    description: 'To configure a special tag for the operations within this rest definition.',
  },
  openApi: {
    title: 'Open Api',
    description: 'To use OpenApi as contract-first with Camel Rest DSL.',
    type: 'object',
    additionalProperties: false,
    properties: {
      description: {
        type: 'string',
        title: 'Description',
        description: 'Sets the description of this node',
      },
      disabled: {
        type: 'boolean',
        title: 'Disabled',
        description:
          'Whether to disable all the REST services from the OpenAPI contract from the route during build time. Once an REST service has been disabled then it cannot be enabled later at runtime.',
      },
      id: {
        type: 'string',
        title: 'Id',
        description: 'Sets the id of this node',
      },
      missingOperation: {
        type: 'string',
        title: 'Missing Operation',
        description:
          'Whether to fail, ignore or return a mock response for OpenAPI operations that are not mapped to a corresponding route.',
        default: 'fail',
        enum: ['fail', 'ignore', 'mock'],
      },
      mockIncludePattern: {
        type: 'string',
        title: 'Mock Include Pattern',
        description:
          'Used for inclusive filtering of mock data from directories. The pattern is using Ant-path style pattern. Multiple patterns can be specified separated by comma.',
        default: 'classpath:camel-mock/**',
      },
      routeId: {
        type: 'string',
        title: 'Route Id',
        description: 'Sets the id of the route',
      },
      specification: {
        type: 'string',
        title: 'Specification',
        description: 'Path to the OpenApi specification file.',
      },
    },
    required: ['specification'],
  },
  securityDefinitions: {
    title: 'Rest Security Definitions',
    description: 'To configure rest security definitions.',
    type: 'object',
    additionalProperties: false,
    properties: {
      apiKey: {
        title: 'Api Key',
        description: 'Rest security basic auth definition',
        type: 'object',
        additionalProperties: false,
        properties: {
          description: {
            type: 'string',
            title: 'Description',
            description: 'A short description for security scheme.',
          },
          inCookie: {
            type: 'boolean',
            title: 'In Cookie',
            description: 'To use a cookie as the location of the API key.',
          },
          inHeader: {
            type: 'boolean',
            title: 'In Header',
            description: 'To use header as the location of the API key.',
          },
          inQuery: {
            type: 'boolean',
            title: 'In Query',
            description: 'To use query parameter as the location of the API key.',
          },
          key: {
            type: 'string',
            title: 'Key',
            description: 'Key used to refer to this security definition',
          },
          name: {
            type: 'string',
            title: 'Name',
            description: 'The name of the header or query parameter to be used.',
          },
        },
        required: ['key', 'name'],
      },
      basicAuth: {
        title: 'Basic Auth',
        description: 'Rest security basic auth definition',
        type: 'object',
        additionalProperties: false,
        properties: {
          description: {
            type: 'string',
            title: 'Description',
            description: 'A short description for security scheme.',
          },
          key: {
            type: 'string',
            title: 'Key',
            description: 'Key used to refer to this security definition',
          },
        },
        required: ['key'],
      },
      bearer: {
        title: 'Bearer Token',
        description: 'Rest security bearer token authentication definition',
        type: 'object',
        additionalProperties: false,
        properties: {
          description: {
            type: 'string',
            title: 'Description',
            description: 'A short description for security scheme.',
          },
          format: {
            type: 'string',
            title: 'Format',
            description: 'A hint to the client to identify how the bearer token is formatted.',
          },
          key: {
            type: 'string',
            title: 'Key',
            description: 'Key used to refer to this security definition',
          },
        },
        required: ['key'],
      },
      mutualTLS: {
        title: 'Mutual TLS',
        description: 'Rest security mutual TLS authentication definition',
        type: 'object',
        additionalProperties: false,
        properties: {
          description: {
            type: 'string',
            title: 'Description',
            description: 'A short description for security scheme.',
          },
          key: {
            type: 'string',
            title: 'Key',
            description: 'Key used to refer to this security definition',
          },
        },
        required: ['key'],
      },
      oauth2: {
        title: 'Oauth2',
        description: 'Rest security OAuth2 definition',
        type: 'object',
        additionalProperties: false,
        properties: {
          authorizationUrl: {
            type: 'string',
            title: 'Authorization Url',
            description:
              'The authorization URL to be used for this flow. This SHOULD be in the form of a URL. Required for implicit and access code flows',
          },
          description: {
            type: 'string',
            title: 'Description',
            description: 'A short description for security scheme.',
          },
          flow: {
            type: 'string',
            title: 'Flow',
            description:
              'The flow used by the OAuth2 security scheme. Valid values are implicit, password, application or accessCode.',
            enum: ['implicit', 'password', 'application', 'clientCredentials', 'accessCode', 'authorizationCode'],
          },
          key: {
            type: 'string',
            title: 'Key',
            description: 'Key used to refer to this security definition',
          },
          refreshUrl: {
            type: 'string',
            title: 'Refresh Url',
            description: 'The URL to be used for obtaining refresh tokens. This MUST be in the form of a URL.',
          },
          scopes: {
            type: 'array',
            title: 'Scopes',
            description: 'The available scopes for an OAuth2 security scheme',
            items: {
              $ref: '#/definitions/org.apache.camel.model.rest.RestPropertyDefinition',
            },
          },
          tokenUrl: {
            type: 'string',
            title: 'Token Url',
            description:
              'The token URL to be used for this flow. This SHOULD be in the form of a URL. Required for password, application, and access code flows.',
          },
        },
        required: ['key'],
      },
      openIdConnect: {
        title: 'Open Id Connect',
        description: 'Rest security OpenID Connect definition',
        type: 'object',
        additionalProperties: false,
        properties: {
          description: {
            type: 'string',
            title: 'Description',
            description: 'A short description for security scheme.',
          },
          key: {
            type: 'string',
            title: 'Key',
            description: 'Key used to refer to this security definition',
          },
          url: {
            type: 'string',
            title: 'Url',
            description: 'OpenId Connect URL to discover OAuth2 configuration values.',
          },
        },
        required: ['key', 'url'],
      },
    },
  },
  securityRequirements: {
    type: 'array',
    title: 'Security Requirements',
    description: 'Sets the security requirement(s) for all endpoints.',
    items: {
      $ref: '#/definitions/org.apache.camel.model.rest.SecurityDefinition',
    },
  },
  delete: {
    type: 'array',
    items: {
      $ref: '#/definitions/org.apache.camel.model.rest.DeleteDefinition',
    },
  },
  get: {
    type: 'array',
    items: {
      $ref: '#/definitions/org.apache.camel.model.rest.GetDefinition',
    },
  },
  head: {
    type: 'array',
    items: {
      $ref: '#/definitions/org.apache.camel.model.rest.HeadDefinition',
    },
  },
  patch: {
    type: 'array',
    items: {
      $ref: '#/definitions/org.apache.camel.model.rest.PatchDefinition',
    },
  },
  post: {
    type: 'array',
    items: {
      $ref: '#/definitions/org.apache.camel.model.rest.PostDefinition',
    },
  },
  put: {
    type: 'array',
    items: {
      $ref: '#/definitions/org.apache.camel.model.rest.PutDefinition',
    },
  },
};
