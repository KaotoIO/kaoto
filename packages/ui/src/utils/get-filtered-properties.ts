import { KaotoSchemaDefinition } from '../models';
import { isDefined } from './is-defined';

/**
 * Extracts the schema recursively containing only the filtered properties.
 */
export function getFilteredProperties(
  properties: KaotoSchemaDefinition['schema']['properties'],
  filter: string,
  omitFields?: string[],
): KaotoSchemaDefinition['schema']['properties'] {
  if (!isDefined(properties)) return {};

  const filteredFormSchema = Object.entries(properties).reduce(
    (acc, [property, definition]) => {
      if (!omitFields?.includes(property)) {
        if (definition['type'] === 'object' && 'properties' in definition) {
          const subFilteredSchema = getFilteredProperties(definition['properties'], filter);
          if (subFilteredSchema && Object.keys(subFilteredSchema).length > 0) {
            acc![property] = { ...definition, properties: subFilteredSchema };
          }
        } else if (property.toLowerCase().includes(filter)) {
          acc![property] = definition;
        }
      }

      return acc;
    },
    {} as KaotoSchemaDefinition['schema']['properties'],
  );

  return filteredFormSchema;
}
