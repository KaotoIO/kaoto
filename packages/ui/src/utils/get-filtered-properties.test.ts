import { restSchemaProperties } from '../stubs/rest-schema-properties';
import { getFilteredProperties } from './get-filtered-properties';

describe('getFilteredProperties()', () => {
  it('should return only the filtered properties', () => {
    const filteredSchema = getFilteredProperties(restSchemaProperties, 'des');
    expect(filteredSchema).toMatchSnapshot();
  });

  it('should return only the un-omitted properties', () => {
    const filteredSchema = getFilteredProperties(restSchemaProperties, '', [
      'get',
      'post',
      'put',
      'delete',
      'patch',
      'patch',
    ]);
    expect(filteredSchema).toMatchSnapshot();
  });
});
