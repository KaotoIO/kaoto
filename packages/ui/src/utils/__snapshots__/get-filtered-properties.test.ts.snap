// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getFilteredProperties() should return only the filtered properties 1`] = `
{
  "description": {
    "description": "Sets the description of this node",
    "title": "Description",
    "type": "string",
  },
  "openApi": {
    "additionalProperties": false,
    "description": "To use OpenApi as contract-first with Camel Rest DSL.",
    "properties": {
      "description": {
        "description": "Sets the description of this node",
        "title": "Description",
        "type": "string",
      },
    },
    "required": [
      "specification",
    ],
    "title": "Open Api",
    "type": "object",
  },
  "securityDefinitions": {
    "additionalProperties": false,
    "description": "To configure rest security definitions.",
    "properties": {
      "apiKey": {
        "additionalProperties": false,
        "description": "Rest security basic auth definition",
        "properties": {
          "description": {
            "description": "A short description for security scheme.",
            "title": "Description",
            "type": "string",
          },
        },
        "required": [
          "key",
          "name",
        ],
        "title": "Api Key",
        "type": "object",
      },
      "basicAuth": {
        "additionalProperties": false,
        "description": "Rest security basic auth definition",
        "properties": {
          "description": {
            "description": "A short description for security scheme.",
            "title": "Description",
            "type": "string",
          },
        },
        "required": [
          "key",
        ],
        "title": "Basic Auth",
        "type": "object",
      },
      "bearer": {
        "additionalProperties": false,
        "description": "Rest security bearer token authentication definition",
        "properties": {
          "description": {
            "description": "A short description for security scheme.",
            "title": "Description",
            "type": "string",
          },
        },
        "required": [
          "key",
        ],
        "title": "Bearer Token",
        "type": "object",
      },
      "mutualTLS": {
        "additionalProperties": false,
        "description": "Rest security mutual TLS authentication definition",
        "properties": {
          "description": {
            "description": "A short description for security scheme.",
            "title": "Description",
            "type": "string",
          },
        },
        "required": [
          "key",
        ],
        "title": "Mutual TLS",
        "type": "object",
      },
      "oauth2": {
        "additionalProperties": false,
        "description": "Rest security OAuth2 definition",
        "properties": {
          "description": {
            "description": "A short description for security scheme.",
            "title": "Description",
            "type": "string",
          },
        },
        "required": [
          "key",
        ],
        "title": "Oauth2",
        "type": "object",
      },
      "openIdConnect": {
        "additionalProperties": false,
        "description": "Rest security OpenID Connect definition",
        "properties": {
          "description": {
            "description": "A short description for security scheme.",
            "title": "Description",
            "type": "string",
          },
        },
        "required": [
          "key",
          "url",
        ],
        "title": "Open Id Connect",
        "type": "object",
      },
    },
    "title": "Rest Security Definitions",
    "type": "object",
  },
}
`;

exports[`getFilteredProperties() should return only the un-omitted properties 1`] = `
{
  "apiDocs": {
    "default": true,
    "description": "Whether to include or exclude this rest operation in API documentation. This option will override what may be configured on a parent level. The default value is true.",
    "title": "Api Docs",
    "type": "boolean",
  },
  "bindingMode": {
    "default": "off",
    "description": "Sets the binding mode to use. This option will override what may be configured on a parent level The default value is auto",
    "enum": [
      "off",
      "auto",
      "json",
      "xml",
      "json_xml",
    ],
    "title": "Binding Mode",
    "type": "string",
  },
  "clientRequestValidation": {
    "default": false,
    "description": "Whether to enable validation of the client request to check: 1) Content-Type header matches what the Rest DSL consumes; returns HTTP Status 415 if validation error. 2) Accept header matches what the Rest DSL produces; returns HTTP Status 406 if validation error. 3) Missing required data (query parameters, HTTP headers, body); returns HTTP Status 400 if validation error. 4) Parsing error of the message body (JSon, XML or Auto binding mode must be enabled); returns HTTP Status 400 if validation error.",
    "title": "Client Request Validation",
    "type": "boolean",
  },
  "consumes": {
    "description": "To define the content type what the REST service consumes (accept as input), such as application/xml or application/json. This option will override what may be configured on a parent level",
    "title": "Consumes",
    "type": "string",
  },
  "description": {
    "description": "Sets the description of this node",
    "title": "Description",
    "type": "string",
  },
  "disabled": {
    "default": false,
    "description": "Whether to disable this REST service from the route during build time. Once an REST service has been disabled then it cannot be enabled later at runtime.",
    "title": "Disabled",
    "type": "boolean",
  },
  "enableCORS": {
    "default": false,
    "description": "Whether to enable CORS headers in the HTTP response. This option will override what may be configured on a parent level The default value is false.",
    "title": "Enable CORS",
    "type": "boolean",
  },
  "enableNoContentResponse": {
    "default": false,
    "description": "Whether to return HTTP 204 with an empty body when a response contains an empty JSON object or XML root object. The default value is false.",
    "title": "Enable No Content Response",
    "type": "boolean",
  },
  "head": {
    "items": {
      "$ref": "#/definitions/org.apache.camel.model.rest.HeadDefinition",
    },
    "type": "array",
  },
  "id": {
    "description": "Sets the id of this node",
    "title": "Id",
    "type": "string",
  },
  "openApi": {
    "additionalProperties": false,
    "description": "To use OpenApi as contract-first with Camel Rest DSL.",
    "properties": {
      "description": {
        "description": "Sets the description of this node",
        "title": "Description",
        "type": "string",
      },
      "disabled": {
        "description": "Whether to disable all the REST services from the OpenAPI contract from the route during build time. Once an REST service has been disabled then it cannot be enabled later at runtime.",
        "title": "Disabled",
        "type": "boolean",
      },
      "id": {
        "description": "Sets the id of this node",
        "title": "Id",
        "type": "string",
      },
      "missingOperation": {
        "default": "fail",
        "description": "Whether to fail, ignore or return a mock response for OpenAPI operations that are not mapped to a corresponding route.",
        "enum": [
          "fail",
          "ignore",
          "mock",
        ],
        "title": "Missing Operation",
        "type": "string",
      },
      "mockIncludePattern": {
        "default": "classpath:camel-mock/**",
        "description": "Used for inclusive filtering of mock data from directories. The pattern is using Ant-path style pattern. Multiple patterns can be specified separated by comma.",
        "title": "Mock Include Pattern",
        "type": "string",
      },
      "routeId": {
        "description": "Sets the id of the route",
        "title": "Route Id",
        "type": "string",
      },
      "specification": {
        "description": "Path to the OpenApi specification file.",
        "title": "Specification",
        "type": "string",
      },
    },
    "required": [
      "specification",
    ],
    "title": "Open Api",
    "type": "object",
  },
  "path": {
    "description": "Path of the rest service, such as /foo",
    "title": "Path",
    "type": "string",
  },
  "produces": {
    "description": "To define the content type what the REST service produces (uses for output), such as application/xml or application/json This option will override what may be configured on a parent level",
    "title": "Produces",
    "type": "string",
  },
  "securityDefinitions": {
    "additionalProperties": false,
    "description": "To configure rest security definitions.",
    "properties": {
      "apiKey": {
        "additionalProperties": false,
        "description": "Rest security basic auth definition",
        "properties": {
          "description": {
            "description": "A short description for security scheme.",
            "title": "Description",
            "type": "string",
          },
          "inCookie": {
            "description": "To use a cookie as the location of the API key.",
            "title": "In Cookie",
            "type": "boolean",
          },
          "inHeader": {
            "description": "To use header as the location of the API key.",
            "title": "In Header",
            "type": "boolean",
          },
          "inQuery": {
            "description": "To use query parameter as the location of the API key.",
            "title": "In Query",
            "type": "boolean",
          },
          "key": {
            "description": "Key used to refer to this security definition",
            "title": "Key",
            "type": "string",
          },
          "name": {
            "description": "The name of the header or query parameter to be used.",
            "title": "Name",
            "type": "string",
          },
        },
        "required": [
          "key",
          "name",
        ],
        "title": "Api Key",
        "type": "object",
      },
      "basicAuth": {
        "additionalProperties": false,
        "description": "Rest security basic auth definition",
        "properties": {
          "description": {
            "description": "A short description for security scheme.",
            "title": "Description",
            "type": "string",
          },
          "key": {
            "description": "Key used to refer to this security definition",
            "title": "Key",
            "type": "string",
          },
        },
        "required": [
          "key",
        ],
        "title": "Basic Auth",
        "type": "object",
      },
      "bearer": {
        "additionalProperties": false,
        "description": "Rest security bearer token authentication definition",
        "properties": {
          "description": {
            "description": "A short description for security scheme.",
            "title": "Description",
            "type": "string",
          },
          "format": {
            "description": "A hint to the client to identify how the bearer token is formatted.",
            "title": "Format",
            "type": "string",
          },
          "key": {
            "description": "Key used to refer to this security definition",
            "title": "Key",
            "type": "string",
          },
        },
        "required": [
          "key",
        ],
        "title": "Bearer Token",
        "type": "object",
      },
      "mutualTLS": {
        "additionalProperties": false,
        "description": "Rest security mutual TLS authentication definition",
        "properties": {
          "description": {
            "description": "A short description for security scheme.",
            "title": "Description",
            "type": "string",
          },
          "key": {
            "description": "Key used to refer to this security definition",
            "title": "Key",
            "type": "string",
          },
        },
        "required": [
          "key",
        ],
        "title": "Mutual TLS",
        "type": "object",
      },
      "oauth2": {
        "additionalProperties": false,
        "description": "Rest security OAuth2 definition",
        "properties": {
          "authorizationUrl": {
            "description": "The authorization URL to be used for this flow. This SHOULD be in the form of a URL. Required for implicit and access code flows",
            "title": "Authorization Url",
            "type": "string",
          },
          "description": {
            "description": "A short description for security scheme.",
            "title": "Description",
            "type": "string",
          },
          "flow": {
            "description": "The flow used by the OAuth2 security scheme. Valid values are implicit, password, application or accessCode.",
            "enum": [
              "implicit",
              "password",
              "application",
              "clientCredentials",
              "accessCode",
              "authorizationCode",
            ],
            "title": "Flow",
            "type": "string",
          },
          "key": {
            "description": "Key used to refer to this security definition",
            "title": "Key",
            "type": "string",
          },
          "refreshUrl": {
            "description": "The URL to be used for obtaining refresh tokens. This MUST be in the form of a URL.",
            "title": "Refresh Url",
            "type": "string",
          },
          "scopes": {
            "description": "The available scopes for an OAuth2 security scheme",
            "items": {
              "$ref": "#/definitions/org.apache.camel.model.rest.RestPropertyDefinition",
            },
            "title": "Scopes",
            "type": "array",
          },
          "tokenUrl": {
            "description": "The token URL to be used for this flow. This SHOULD be in the form of a URL. Required for password, application, and access code flows.",
            "title": "Token Url",
            "type": "string",
          },
        },
        "required": [
          "key",
        ],
        "title": "Oauth2",
        "type": "object",
      },
      "openIdConnect": {
        "additionalProperties": false,
        "description": "Rest security OpenID Connect definition",
        "properties": {
          "description": {
            "description": "A short description for security scheme.",
            "title": "Description",
            "type": "string",
          },
          "key": {
            "description": "Key used to refer to this security definition",
            "title": "Key",
            "type": "string",
          },
          "url": {
            "description": "OpenId Connect URL to discover OAuth2 configuration values.",
            "title": "Url",
            "type": "string",
          },
        },
        "required": [
          "key",
          "url",
        ],
        "title": "Open Id Connect",
        "type": "object",
      },
    },
    "title": "Rest Security Definitions",
    "type": "object",
  },
  "securityRequirements": {
    "description": "Sets the security requirement(s) for all endpoints.",
    "items": {
      "$ref": "#/definitions/org.apache.camel.model.rest.SecurityDefinition",
    },
    "title": "Security Requirements",
    "type": "array",
  },
  "skipBindingOnErrorCode": {
    "default": false,
    "description": "Whether to skip binding on output if there is a custom HTTP error code header. This allows to build custom error messages that do not bind to json / xml etc, as success messages otherwise will do. This option will override what may be configured on a parent level",
    "title": "Skip Binding On Error Code",
    "type": "boolean",
  },
  "tag": {
    "description": "To configure a special tag for the operations within this rest definition.",
    "title": "Tag",
    "type": "string",
  },
}
`;
